src/dioptra/restapi/errors.py

```diff
@@ -20,87 +20,363 @@
 """
 from __future__ import annotations
 
+import http
+import typing
+from io import StringIO
+
+import structlog
+from flask import request
 from flask_restx import Api
+from structlog.stdlib import BoundLogger
+
+LOGGER: BoundLogger = structlog.stdlib.get_logger()
+
+"""
+Helper function to add attribute name/value pairs to StringIO instance as an error
+message.
+Args:
+    buffer: The StringIO instance into which the attribute name/value pairs are to be
+    added.
+    attributes: the list of attribute value pairs to add to the buffer.
+"""
+
+
+def add_attribute_values(buffer: StringIO, **kwargs: typing.Any) -> None:
+    index = 0
+    length = len(kwargs)
+    for key, value in kwargs.items():
+        if index != 0:
+            buffer.write(", ")
+            if index == length - 1:
+                buffer.write("and ")
+        buffer.write(f"{key} having value ({value})")
+        index += 1
+
+
+class DioptraError(Exception):
+    """
+    Generic Dioptra Error.
+    Args:
+        message: An error specific message to display that provide context for why the
+        error was raised.
+    """
+
+    def __init__(self, message: str):
+        self.message: str = message
+
+    def to_message(self) -> str:
+        if self.__cause__ is not None:
+            if isinstance(self.__cause__, DioptraError):
+                self.message = f"{self.message} Cause: {self.__cause__.to_message()}"
+            else:
+                self.message = f"{self.message} Cause: {self.__cause__}"
+        return self.message
```

Comment:
This implementation can append to the message on each call:

```
>>> e = errors.EntityDoesNotExistError("job", key1="value1")
>>> e.__cause__ = ValueError("bad value")
>>> e.to_message()
'Failed to locate job with key1 having value (value1). Cause: bad value'
>>> e.to_message()
'Failed to locate job with key1 having value (value1). Cause: bad value Cause: bad value'
>>> e.to_message()
'Failed to locate job with key1 having value (value1). Cause: bad value Cause: bad value Cause: bad value'
```

I think it should be safe to call .to_message() more than once without corrupting the message.

Reply:
Agree on this point. From what I can tell, the `.to_message()` method is only used to populate `log.debug()` in the registered error handlers. If that case, why not this?

```python
def to_message(self) -> str:
    if self.__cause__ is not None:
        if isinstance(self.__cause__, DioptraError):
            return f"{self.message} Cause: {self.__cause__.to_message()}"
  
        return f"{self.message} Cause: {self.__cause__}"

    return self.message
```